impl/src/main/java/com/okta/sdk/impl/client/DefaultClientBuilder.java

```diff
@@ -307,18 +325,63 @@ public Client build() {
             this.cacheManager = cacheManagerBuilder.build();
         }
 
-        if (this.clientConfig.getClientCredentialsResolver() == null && this.clientCredentials != null) {
-            this.clientConfig.setClientCredentialsResolver(new DefaultClientCredentialsResolver(this.clientCredentials));
+        if (isOAuth2Flow()) {
+            this.clientConfig.setClientCredentialsResolver(new OAuth2ClientCredentialsResolver(this.clientCredentials));
         }
-        else if (this.clientConfig.getClientCredentialsResolver() == null) {
-            this.clientConfig.setClientCredentialsResolver(new DefaultClientCredentialsResolver(clientConfig));
+        else {
+            if (this.clientConfig.getClientCredentialsResolver() == null && this.clientCredentials != null) {
+                this.clientConfig.setClientCredentialsResolver(new DefaultClientCredentialsResolver(this.clientCredentials));
+            } else if (this.clientConfig.getClientCredentialsResolver() == null) {
+                this.clientConfig.setClientCredentialsResolver(new DefaultClientCredentialsResolver(this.clientConfig));
+            }
         }
 
         if (this.clientConfig.getBaseUrlResolver() == null) {
             ConfigurationValidator.assertOrgUrl(this.clientConfig.getBaseUrl(), allowNonHttpsForTesting);
             this.clientConfig.setBaseUrlResolver(new DefaultBaseUrlResolver(this.clientConfig.getBaseUrl()));
         }
 
+        // OAuth2
+        if (isOAuth2Flow()) {
+            this.clientConfig.setAuthenticationScheme(AuthenticationScheme.OAUTH2);
+            accessTokenRetrieverService = new AccessTokenRetrieverServiceImpl(clientConfig);
+
+            // Get access token asynchronously
+            CompletableFuture<OAuth2AccessToken> completableFuture = CompletableFuture.supplyAsync(() -> {
```

Comment:
This likely should NOT be done in the builder class but instead, in the appropriate client credentials (or related) object, this is going to be difficult to test this way.

Reply:
Moved it out of here.